### ISSUE TITLE
Using Multiple GPUs/CPUs (CNTK) with C++ cntk

### ISSUE THREAD

**Issue by Ivashar** [Microsoft/CNTK#3551]
**I use the CNTK library (C++) for learning LSTM on a single CPU or GPU. I do not understand how to change my code and run training on multiple GPUs (CPU). My current code is:**

`labels = InputVariable({ numOutputClasses }, DataType::Float, L"labels");
trainingLoss = CrossEntropyWithSoftmax(lstmModel, labels, 
L"lossFunction");
prediction   = CrossEntropyWithSoftmax(lstmModel, labels, 
L"classificationError");

//create learner 
paramLearner = AdamLearner(lstmModel->Parameters(),
                          learningRate,
                                momentumSchedule, 
                                false);

//create trainer
trainer = CreateTrainer(lstmModel, trainingLoss, prediction, vector<LearnerPtr>({ paramLearner }));

sampleShape = { inputDim };
labelsShape = { numOutputClasses };

classifierOutputVar = lstmModel->Output();
unordered_map<Variable, ValuePtr> argumentsOut;
double trainLossValue;

// run train
for (size_t i = 1; i <= countEpoch; ++i)
{
    cout << "Epoch " << i << ":" << countEpoch << endl;

    for (int k = 0; k < inputData.size(); ++k) 
    {
        argumentsOut = { { classifierOutputVar, outputValue },
                         { prediction, predictionErrorValue } };

        featuresValue = Value::Create(sampleShape, inputData.at(k),  device);
        labelValue    = Value::Create(labelsShape, labelsData.at(k), device);
        argumentsIn = { { features, featuresValue }, { labels, labelValue } };

        trainer->TrainMinibatch(argumentsIn, true, argumentsOut, device);
        argumentsIn.clear();

        trainLossValue = trainer->PreviousMinibatchLossAverage();
        cout << "\tBatch " << k + 1 << ":" << inputData.size() << "\ttrainLossValueBatch: " << trainLossValue << endl;
    }

    saveModel(path);        
}`

**inputData is a vector vector <vector <vector <float >>> containing several MFCC vectors divided by packs. This is done because there is a lot of data and the launch leads to a crash (memory overflow and operating system failure). Why does the library allow this ??? How to track this ??? But the question is not about that.**

Attempts to figure it out on their own have failed:
`auto sync = MPICommunicator();

auto numWorkers = sync->Workers().size();
auto workerRank = sync->CurrentWorker().m_globalRank;

labels = InputVariable({ numOutputClasses }, DataType::Float, L"labels");
trainingLoss = CrossEntropyWithSoftmax(lstmModel, labels, L"lossFunction");
prediction = ClassificationError(lstmModel, labels, L"classificationError");

paramLearner = FSAdaGradLearner(lstmModel->Parameters(),
                                learningRate,
                                momentumSchedule,
                                false);

DistributedLearnerPtr distributedLearner =
    CreateDataParallelDistributedLearner(MPICommunicator(), paramLearner, 0);

trainer = CreateTrainer(lstmModel, trainingLoss, prediction, { distributedLearner });`

**It is not clear how to specify a list of GPU (CPU) to run ...

How to run training using the inputData container on multiple CPU / GPU (or PC) ??? I consider examples, but it is not quite clear how to organize the calculation.
It is known that the program must be run as mpiexec .... (arguments).
Please, help !!!**
